feat(friends): Clear conversation after person is removed from friend list

drepram opened the pull request:
<!--  Thanks for sending a pull request!-->

### What this PR does 📖
- title

### Which issue(s) this PR fixes 🔨
<img width="948" alt="image" src="https://user-images.githubusercontent.com/34530026/204032105-42ea0986-f3d8-4b06-b747-d7f00e7aa310.png">
on #395 

### Special notes for reviewers 🗒️

Lucas and I found a bug relating to this. Sometimes, only one person will have a clear chat while the other person will retain the chat history. I am still looking into this, and experimenting with MultiPassEventKind to see if we can mitigate this defect.

### Additional comments 🎤

drepram committed:
feat(friends): wip clear chat when friend removed

drepram committed:
feat(friends): wip clear chat when friend removed

drepram committed:
feat(friends): succesfully clear conversation

drepram committed:
feat(friends): succesfully clear conversation

drepram commented:
After running `cargo clean`, I can't seem to reproduce my own error. Please let me know if it does occur on your machine :)

drepram committed:
chore(rust): patch warnings and add remove friend on contextmenu

drepram committed:
chore(rust): patch warnings and add remove friend on contextmenu

phillsatellite commented:
Having issues when trying to compile, asked Sara to verify and she also had same issues

<img width="1531" alt="Screenshot 2022-11-25 at 5 53 41 PM" src="https://user-images.githubusercontent.com/93608357/204061513-4e4d9b24-0556-4bfb-be5f-2b39852a30d7.png">

drepram committed:
chore(git): merge dev into branch

drepram committed:
chore(friends): update icons

drepram committed:
chore(git): fix merge conflict, again

drepram committed:
chore(rust): remove trailing white space

drepram committed:
chore(git): reorder code for rustfmt

stavares843 commented:
still happens

phillsatellite commented:
Hi Andre! So Sara wasn't able to compile but I was and only found one minor issue when testing, 

It seems like when User removes friend, the chat messages will be removed but it looks like User will be left with an empty chat

https://user-images.githubusercontent.com/93608357/204342450-284c367d-4cdb-466e-b0a3-2593fbbe43f3.mov

InfamousVague commented:
I think the problem Phil is experiencing can be fixed by removing the active_chat and setting it back to the default value in state 
<img width="380" alt="image" src="https://user-images.githubusercontent.com/1770198/204369033-131a9e15-c369-47be-a685-471f18b4f49a.png">

Setting this to `None` may solve the issue. 

Maybe something like 

```rs
Actions::ClearChat => PersistedState {
                current_chat: None,
                all_chats: self.all_chats.clone(),
                favorites: self.favorites.clone(),
                hide_sidebar: self.hide_sidebar,
                total_unreads: self.total_unreads,
            },
            ```
 ```rs
 pub enum Actions {
    AddRemoveConversations(HashMap<Uuid, ConversationInfo>),
    ChatWith(ConversationInfo),
    UpdateConversation(ConversationInfo),
    UpdateFavorites(HashSet<Uuid>),
    HideSidebar(bool),
    ClearChat,
    // SendNotification(String, String, Sounds),
}
```

dariusc93 commented:
@drepram could you fix the merge conflicts

drepram committed:
chore(git): fix merge conflict

dariusc93 committed:
fix: Remove write locks

drepram committed:
Merge branch 'dev' into drepram/clear-conversation

drepram committed:
chore(rust): resolve errors

drepram committed:
feat(friends): Clear conversation via context menu, and friend list

drepram commented:
Hi, based upon the input I've received I've decided to make it into two options in the context menu:

<img width="472" alt="image" src="https://user-images.githubusercontent.com/34530026/204596403-f5deeebb-d675-45f9-8059-3d5eb21b46a4.png">

`Clear Chat` will only clear **that** chat, it will not remove the friend from the friend list. If we do click `Remove Friend` however, it will clear chat AND remove the friend from the list; this behavior is identical to the friend page.

Should these new options be in the chat ContextMenu, or should they be here on the sidebar instead:

<img width="326" alt="image" src="https://user-images.githubusercontent.com/34530026/204597918-454df643-0d50-42be-b240-d32aeb39f2af.png">

Let me know what y'all think. Thanks for reading and have a good day!

dariusc93 commented on src/components/main/friends/friend/mod.rs:
```diff
@@ -27,7 +27,6 @@ pub struct Props<'a> {
 #[allow(non_snake_case)]
 pub fn Friend<'a>(cx: Scope<'a, Props>) -> Element<'a> {
     log::debug!("rendering Friend");
-    let state = use_atom_ref(&cx, STATE);
```
Instead of removing it and using it in 2 locations you could simply revert this and clone the state by doing `let stat2 = state.clone()`, etc., and use `state2` in the section location.

drepram replied:
Done! I also made `remove_friend` function to reduce redundancy.

dariusc93 reviewed (approved):
Everything looks good so far from a glance over. Left a comment but nothing blocking.

drepram committed:
Merge branch 'dev' into drepram/clear-conversation

drepram committed:
feat(friend): resolve bug, remove redundancy

phillsatellite commented:
I found one other minor issue when re testing the latest changes, sent to Andre in Discord as well 👍 
https://user-images.githubusercontent.com/93608357/204878491-bc1095a4-0965-46e0-add5-4ecdeebf797c.mov

drepram committed:
feat(friends): create chat when friend is added, add raygun to props

drepram committed:
Merge branch 'dev' into drepram/clear-conversation

phillsatellite commented:
Hi Andre! I think one minor change is needed I'm sorry 😂 😢  I got this error on multiple devices when trying to do cargo run 

<img width="1001" alt="Screenshot 2022-12-01 at 2 25 34 PM" src="https://user-images.githubusercontent.com/93608357/205142026-92f011b1-1610-4f98-b13c-fe1eb054d4ce.png">

sdwoodbury reviewed (changes requested):
I think this PR can be abandoned. PR's [497](https://github.com/Satellite-im/Uplink/pull/497) and [502](https://github.com/Satellite-im/Uplink/pull/502) have been merged in. They add the capability to remove a conversation when the friend is deleted. The feature isn't yet implemented but I think that now all it will take is deleting a conversation from RayGun. I'm not sure how to do that,but when the `RayGunEventKind::ConversationDeleted` is emitted, the conversation will be removed from the sidebar.

drepram committed:
chore(rust): patch missing variable declaration

drepram committed:
chore(rust): patch formatting warnings

drepram committed:
chore(git): fix merge conflict

drepram committed:
Merge branch 'dev' into drepram/clear-conversation

drepram commented:
Hi Stu, I rechecked my code and I saw that Phill and Darius's comment was correct. Everything looks good so far and only one minor issue needs to be fixed. As it turns out, there was a missing variable declaration on src/components/main/files/mod.rs; I patched it.

I checked in with Joe, and it does work on his machine. You see, my PR was not only about https://github.com/Satellite-im/Uplink/pull/395; I was trying to kill two birds with one stone and my last commits were also in the direction of solving https://github.com/Satellite-im/Uplink/issues/404.

Please take another look, it seems that I've implemented the feature these issues brought about. Thanks!

dariusc93 commented:
> I think this PR can be abandoned. PR's [497](https://github.com/Satellite-im/Uplink/pull/497) and [502](https://github.com/Satellite-im/Uplink/pull/502) have been merged in. They add the capability to remove a conversation when the friend is deleted. The feature isn't yet implemented but I think that now all it will take is deleting a conversation from RayGun. I'm not sure how to do that,but when the `RayGunEventKind::ConversationDeleted` is emitted, the conversation will be removed from the sidebar.

This can be done by listening on `MultiPassEventKind::FriendRemoved` and iterating over the conversation list to find the other did (make sure its to check that the conversation type is `ConversationType::Direct` to be safe) then take the conversation id and use `RayGun::delete` with that id. Once `RayGunEventKind::ConversationDeleted` is emitted it can be removed from the sidebar there.

drepram committed:
Merge branch 'dev' into drepram/clear-conversation

drepram committed:
chore(rust): correct rust formatting

lgmarchi committed:
Fix: Fix build errors

drepram committed:
chore(chat): WIP ChatWith implementation

drepram committed:
feat(build): fix build error

stavares843 commented on src/components/main/friends/friend/mod.rs:
```diff
@@ -1,14 +1,18 @@
     skeletons::{inline::InlineSkeleton, pfp::PFPSkeleton},
 };
-use warp::crypto::DID;
+use warp::{crypto::DID, raygun::Conversation};
```
<img width="840" alt="Captura de ecrã 2022-12-05, às 18 57 14" src="https://user-images.githubusercontent.com/29093946/205720093-9cefb008-5fb9-4206-a41c-d755ab97c6dc.png">

drepram replied:
I can resolve the other warnings, but what about the `non_snake_case` and the props' "`unused_variables`"? May I ignore those?

stavares843 replied:
@sdwoodbury what do you think?

dariusc93 replied:
@drepram you could ignore those by doing `#[allow(non_snake_case)]` on to of `Friends` and probably the same for the unused variables. Can also remove the mut there if its not needed for `rg`

drepram replied:
You are right!

sdwoodbury commented on src/components/main/compose/msg/mod.rs:
```diff
@@ -235,6 +236,58 @@ pub fn Msg<'a>(cx: Scope<'a, Props>) -> Element<'a> {
                                 text: String::from("Reply"),
                                 icon: Shape::ArrowUturnLeft,
+                            },
+                            ContextItem {
```
it looks like this is deleting the conversation? I don't think we want to do that. I think we want to be able to remove a conversation from the sidebar, but then get that conversation back by clicking on the chat icon from the Friends list.

drepram replied:
Resolved by the latest commit

drepram replied:
We have a hide chat feature here:

<img width="224" alt="image" src="https://user-images.githubusercontent.com/34530026/206525016-6097c53e-9a6a-47d6-8030-1989470355d7.png">


And a clear chat feature:

<img width="291" alt="image" src="https://user-images.githubusercontent.com/34530026/206525119-ea2e8c2b-fe4b-4537-85df-e276b4e968de.png">

They are separate, and I think there is a distinction between them. Should I remove the Clear Chat feature altogether? Or let it complement the features of the chat? Thanks!

sdwoodbury commented on src/components/main/compose/msg/mod.rs:
```diff
@@ -235,6 +236,58 @@ pub fn Msg<'a>(cx: Scope<'a, Props>) -> Element<'a> {
+                                        async move {
+                                            match rg.delete(conversation_id, None).await {
+                                                Ok(_) => {
+                                                    state.write().dispatch(Actions::RemoveConversation(conversation_id));
```
i believe that per @dariusc93 's comment, this dispatch should happen after line 282.

drepram replied:
You are correct.

sdwoodbury commented on src/components/main/compose/msg/mod.rs:
```diff
@@ -235,6 +236,58 @@ pub fn Msg<'a>(cx: Scope<'a, Props>) -> Element<'a> {
+                                    });
+                                    let mut multipass = cx.props.account.clone();
+                                    let did_to_remove = cx.props.sender.clone();
+                                    match multipass.remove_friend(&did_to_remove) {
```
this generates an event. The remote side should listen for that event and remove the friend and chat accordingly.

drepram replied:
Resolved by the latest commit

sdwoodbury commented on src/components/main/friends/friend/mod.rs:
```diff
@@ -92,12 +109,35 @@ pub fn Friend<'a>(cx: Scope<'a, Props>) -> Element<'a> {
-                                    Ok(_) => {}
-                                    Err(_) => {
-                                        log::debug!("error removing friend");
+                                let local_state = use_atom_ref(&cx, STATE).clone();
```
this will break the UI. the same hooks must be created across all invocations of an Element. creating a hook in a callback, loop, or conditional breaks that rule.

drepram replied:
Resolved by the latest commit

sdwoodbury commented on src/components/main/friends/friend/mod.rs:
```diff
@@ -92,12 +109,35 @@ pub fn Friend<'a>(cx: Scope<'a, Props>) -> Element<'a> {
-                                        log::debug!("error removing friend");
+                                let local_state = use_atom_ref(&cx, STATE).clone();
+                                let rg = cx.props.messaging.clone();
+                                let current_chat_exist = local_state.read().selected_chat.clone();
```
remove this

sdwoodbury replied:
line 114

drepram replied:
Resolved by the latest commit

sdwoodbury commented on src/components/main/friends/friend/mod.rs:
```diff
@@ -92,12 +109,35 @@ pub fn Friend<'a>(cx: Scope<'a, Props>) -> Element<'a> {
+                                let local_state = use_atom_ref(&cx, STATE).clone();
+                                let rg = cx.props.messaging.clone();
+                                let current_chat_exist = local_state.read().selected_chat.clone();
+                                match current_chat_exist {
```
match on `local_state.read().selected_chat.clone()`

drepram replied:
Great point

sdwoodbury commented on src/components/main/friends/friend/mod.rs:
```diff
@@ -92,12 +109,35 @@ pub fn Friend<'a>(cx: Scope<'a, Props>) -> Element<'a> {
+                                let current_chat_exist = local_state.read().selected_chat.clone();
+                                match current_chat_exist {
+                                    Some(_) => {
+                                        let current_chat = local_state.read().selected_chat.and_then(|x| local_state.read().active_chats.get(&x).cloned());
```
remove this

sdwoodbury replied:
line 117

drepram replied:
Resolved by the latest commit

sdwoodbury commented on src/components/main/friends/friend/mod.rs:
```diff
@@ -92,12 +109,35 @@ pub fn Friend<'a>(cx: Scope<'a, Props>) -> Element<'a> {
+                                        let current_chat = local_state.read().selected_chat.and_then(|x| local_state.read().active_chats.get(&x).cloned());
+                                        let current_chat_condition = match current_chat {
+                                            Some(c) => c,
+                                            None => return,
```
returning here will mean you don't delete the friend. the friend should probably be deleted in any case.

drepram replied:
Resolved by the latest commit

phillsatellite commented:
Hi Andre! I also found a small little issue when testing, if both of the users do not send a message in the chat when one of them removes the other the chat will still appear in sidebar

This is showing the bug, the Main Account sent a message to "Second Friend". Instead of sending message back Second Friend removes Main Account but chat still appear on both users side

https://user-images.githubusercontent.com/93608357/205724627-f4c3e8ca-2328-49dd-9c06-5a4f5089eece.mov

Here shows chat successfully being removed after both users have sent a message  

https://user-images.githubusercontent.com/93608357/205724898-0ca0fb4b-24f3-4d36-b7ed-c238afc47afc.mov

dariusc93 commented on src/components/reusable/nav/mod.rs:
```diff
@@ -107,11 +110,14 @@ pub fn Nav(cx: Scope<Props>) -> Element {
                         }
+                        log::debug!("creating chat");
+                        let _result =
+                            warp::async_block_in_place_uncheck(rg.create_conversation(&did));
```
Since this is inside of a use_future, you can remove the `warp::async_block_in_place_uncheck` and just do `rg.create_conversation(&did).await`. I would also suggest checking for any errors (similar to what was being done previously) and log them.

drepram replied:
Great point!

drepram committed:
Merge branch 'dev' into drepram/clear-conversation

lgmarchi committed:
Merge branch 'dev' of https://github.com/Satellite-im/Uplink into drepram/clear-conversation

lgmarchi committed:
refactor: Hide conversation if friend is removed, but not delete it

lgmarchi committed:
Merge branch 'dev' of https://github.com/Satellite-im/Uplink into drepram/clear-conversation

lgmarchi reviewed (approved):
Not updating in real time, but if we go out chats section and back, or it hide chat, or it show chats again as video below. 

https://user-images.githubusercontent.com/63157656/206460716-30ff279f-7bbf-44d5-ba4b-e8cd930365bf.mov

drepram committed:
Merge branch 'dev' into drepram/clear-conversation

drepram committed:
chore(rust): formatting

drepram committed:
chore(rust): clippy fix

drepram committed:
chore(git): fix merge conflict

drepram committed:
chore(rust): fix clippy

sdwoodbury committed:
fix (chats): handle the FriendRemoved event from MultiPass and overcome RayGun shortcoming

This seems the best we can do for now to remove conversations when a friend is removed

sdwoodbury committed:
properly remove conversations and simplify logic

sdwoodbury committed:
fix(clippy)

sdwoodbury committed:
fix (clear conversation): chat is now cleared for the sender

sdwoodbury committed:
refactor: remove unnecessary variable

sdwoodbury committed:
refactor: remove unneeded variables

sdwoodbury committed:
fix(fmt)

sdwoodbury committed:
fix (clear chat): should hide the chat. renamed UI too
